main.cpp

```diff
@@ -0,0 +1,247 @@
+/* Stellar-CSO
+   A tool for converting ISO/XISO files to CSO (Compressed ISO)
+
+   Copyright 2018 David O'Rourke <david.orourke@gmail.com>
+   Copyright 2022 MakeMHz LLC <contact@makemhz.com>
+   Based on ciso from https://github.com/jamie/ciso
+*/
+
+#include <lz4frame.h>
+#include <lz4hc.h>
+
+#include <iostream>
+#include <ios>
+#include <filesystem>
+#include <fstream>
+#include <string>
+#include <vector>
+
+const uint32_t FATX_SIZE_LIMIT = 0xFFBF6000;
+
+const std::string CISO_MAGIC = "CISO";
+const uint32_t CISO_HEADER_SIZE = 0x18;
+const uint32_t CISO_BLOCK_SIZE = 0x800;
+const uint32_t CISO_PLAIN_BLOCK = 0x80000000;
+
+size_t image_offset = 0;
+
+void detect_iso_type(std::ifstream& f) {
+    try {
+	f.seekg(0x18310000, std::ios::beg);
+    } catch (std::exception const& e) {
+	std::cout << "File not big enough to be a Redump style image." << std::endl;
+    }
+    if (f.tellg() == 0x18310000) {
+	std::string buf(20, '\0');
+	f.read(&buf[0], 20);
+	if (buf == "MICROSOFT*XBOX*MEDIA") {
+	    image_offset = 0x18300000;
+	    return;
+	}
+    }
+    if (f.bad() || f.fail() || f.eof()) {
+        f.clear();
+    }
+
+    try {
+	f.seekg(0x10000, std::ios::beg);
+    } catch (std::exception &e) {
+	std::cout << "File not big enough to be a raw XDVDFS image." << std::endl;
+    }
+    if (f.tellg() == 0x10000) {
+	std::string buf(20, '\0');
+	f.read(&buf[0], 20);
+	if (buf == "MICROSOFT*XBOX*MEDIA") {
+	    image_offset = 0;
+	    return;
+	} else {
+            std::cout << buf << std::endl;
+        }
+    }
+    std::cerr << "Could not detect ISO type."<< std::endl;
+    exit(1);
+}
+
+class ciso {
+private:
+    std::string magic;
+    uint32_t ver;
+    uint32_t block_size;
+    uint64_t total_bytes;
+    uint32_t total_blocks;
+    uint32_t align;
+public:
+    ciso (uint64_t total_bytes) : magic(CISO_MAGIC), ver(2), block_size(CISO_BLOCK_SIZE),
+				  total_bytes(total_bytes), align(2) {
+	total_blocks = total_bytes/CISO_BLOCK_SIZE;
+    }
+
+    ciso (std::ifstream &f) {
+	f.seekg(0, std::ios::end);
+	uint32_t file_size = (int32_t)f.tellg() - image_offset;
```

Comment:
This seems to be causing issues, should (probably) be `uint64_t` instead?

Reply:
That brings us closer at least; now redump input results in two files with reasonable sizes. Output doesn't match the python script past the header though...